V1 (#4)
ImmutableWebApps/immutablewebapps.github.io: README.md

## Concepts

The following concepts define the core requirements for an _Immutable Web Application_. They are framework and infrastructure agnostic.

### Static assets are independent of the web application environment(s)

Static assets are the files (javascript, css, images) that are generated from a build of a web application codebase. They become _immutable_ when they do not contain anything environment-specific and they are published to a unique, independent location.

#### Static assets must not contain anything that is environment-specific

All of the leading application frameworks ([Angular CLI](https://github.com/angular/angular-cli/wiki/stories-application-environments), [Create React App](https://cli.vuejs.org/guide/mode-and-env.html#using-env-variables-in-client-side-code), [Ember CLI](https://ember-cli.com/user-guide/#Environments), [Vue CLI 3](https://cli.vuejs.org/guide/mode-and-env.html#using-env-variables-in-client-side-code)) recommend defining environment _values_ at _compile time_. This practice requires that the static assets are generated for each environment and regenerated for any change to an environment.

> _Immutable Web Applications_ reference environment _variables_ that are  defined on the global scope and that are set to values in the `index.html` that unique to the deployment:

Reviewer: that are unique

Author: ```suggestion
_Immutable Web Applications_ reference environment _variables_ that are  defined on the global scope and that are set to values in the `index.html` that are unique to the deployment:
```

```diff
  import { Injectable } from '@angular/core';
  import { HttpClient } from '@angular/common/http';

- // remove environment constants that are defined in the codebase or at compile time
- import { environment } from './environments/environment';

  @Injectable()
  export class DataService {
    configUrl = ;

    constructor(
        private http: HttpClient
        private
        ) { }

    getData() {
-      return this.http.get(`${environment.API}/assets/data.json`);
+      // use globally scoped environment variables that are unique to the deployment

> +      return this.http.get(`${window.env.API}/assets/data.json`);

Reviewer 2: Just an FYI that this will likely get some blowback from Angular people.  Since Angular is built to be somewhat platform agnostic, referencing window directly in anything but one of the top level components or app bootstrapper will make some people twitch.

Author: I will work on a different example that uses `environment.js` for v2

    }
  }
```

#### Static assets must be hosted at locations that are unique and _independent of the web application environment(s)_.

> Static assets that do not contain anything environment-specific can be built once, published to once to a unique location, and then used in multiple environments of the web application.

Reviewer 2: ```suggestion
Static assets that do not contain anything environment-specific can be built once, published to a unique location, and then used in multiple environments of the web application.
```

Author: @ryancarroll I am going to accept your suggestion, but it is not a typo and was included intentionally. But I am always in favor of less words over more words... and they don't add much value.

These static assets share the same qualities as hosted javascript libraries on content delivery networks (CDN) ([Google Hosted Libraries](https://developers.google.com/speed/libraries/), [cdnjs](https://cdnjs.com/), [jsDelivr](https://www.jsdelivr.com/), [UNPKG](https://unpkg.com/)):

<code>https:/<span></span>/<span style="font-weight: bold">ajax.googleapis.c<span></span>om</span>/ajax/libs/jquery/<span style="font-weight: bold">3.3.1</span>/jquery.min.js</code>

The location of the jquery library above, referenced by innumerable web applications, is both independent of the applications and uniquely versioned.

<code>https:/<span></span>/<span style="font-weight: bold">assets.myapp.c<span></span>om</span>/apps/<span style="font-weight: bold">1.0.2</span>/main.js</code>

> Similarly, the location of the example the web application javascript file is unique and hosted at a location that is dedicated to static assets. The static asset web server is a repository for versions of the web application.

Reviewer 2: This reads a bit weird...  Maybe this instead...

```suggestion
Similarly, the location of the web application javascript files are unique and hosted at a location that is dedicated to static assets. The static asset web server is a repository for versions of the web application.
```
Might also make sense to move the URL example below this paragraph.

Author: 👍

#### Configure the static assets for long-term caching

Static assets that do not contain anything environment-specific and  are hosted at a unique and permanent location may be configured to be [cached by the browser (almost) indefinitely](https://developer.mozilla.org/en-US/docs/Web/HTTP/Headers/Cache-Control#Caching_static_assets):

 `cache-control: public, max-age=31536000, immutable`

### `index.html` is deployable configuration

 The HTML document of a single-page application (often `index.html`) is ___not___ static. It varies from environment to environment, and deploy to deploy. The HTML document is a composition of the environment-specific configuration and _immutable_ static assets that define the web application.

#### `index.html` contains fully-qualified references to the static assets

```diff
- <!-- not unique, not independent -->
- <script src="main.js" type="text/javascript"></script>

+ <!-- permanent, reuseable -->

> + <script src="https://assets.myapp.com/apps/bae1407/main.js" type="text/javascript"></script>

Reviewer 2: What is bae1407??  Might be better to carry through the same example asset name/version that you refer to earlier in the doc.

Generally the codebase is a source controlled code repository that is capable of building, versioning, and publishing static assets to a static web server.

_An example of this might be:_

An __Angular__ project hosted in a __GitHub__ repo. The repo is integrated with __TravisCI__ to build and version assets when commits are pushed to the master branch. Versioned assets are published to a unique location in an __AWS S3 bucket__.

### Releasing

The `index.html` files are managed independently of the codebase and they serve as a manifest for each environment. They should be considered configuration files and managed accordingly. Additionally, There needs to be a mechanism for modifying or replacing the `index.html` in each web application environment. The act of changing an `index.html` is effectively a deployment.

_An example of this might be:_

> A set of `index.html` files, one for each environment, hosted in a __Github__ repo. The repo is integrated with __TravisCI__ to publish an `index.html` file when it is modified to an __AWS S3 bucket__. There is an `index.html` and S3 bucket dedicated to each web application environment.

Reviewer 2: This reads as though there needs to be a separate S3 bucket for every environment.  That isn't necessarily required, correct?  I understand you are just trying to lay out an example though, just wondering if you need to be that prescriptive with the example.

### Infrastructure

The infrastructure to support _Immutable Web Apps_ is composed of three parts:

- A static web server to host the web application
- A static web server to host the static assets

> - A content delivery network to make optimize the delivery

Reviewer 2: ```suggestion
- A content delivery network to optimize the delivery of the web application and static assets
```

### Atomic live releases

A live releases of an _Immutable Web App_ is the act of publishing a single `index.html` file. The deployment is instantaneous and all assets are available immediately without any risk of cache being corrupted at the time of the release.

Rollbacks are as risky as deployments and often riskier. For _Immutable Web Apps_, the same qualities of a deploy apply to a rollback. Notably, in the event of a rollback most browsers will still have the previous assets cached.

In the unlikely event that a browser attempts to load a stale version of `index.html`, all of the assets of the previous version will still be available and uncorrupted.

### Simplified caching

Managing `cache-control` headers can be intimidating, especially when the web application infrastructure leverages public caches like the ones used by CDNs. The two simplest concepts in caching are: "Always cached", and "Never cached". _Immutable Web Apps_ embraces these concepts completely separating and the code that can be "Always cached" from configuration that is "Never cached".

### Simplified routing

> The leading application frameworks do not separate the location of static assets from the `index.html` in their [deployment recommendations](https://angular.io/guide/deployment#server-configuration). Instead, they recommend adding routing rules to the web server that returns `index.html` for all paths that do not resolve to a physical file. The implementation of these routing rules may differ from between web servers and mistakes often result in paths that resolve to the wrong resource.

Reviewer 2: ```suggestion
The leading application frameworks do not separate the location of static assets from the `index.html` in their [deployment recommendations](https://angular.io/guide/deployment#server-configuration). Instead, they recommend adding routing rules to the web server that returns `index.html` for all paths that do not resolve to a physical file. The implementation of these routing rules may differ between web servers and mistakes often result in paths that resolve to the wrong resource.
```
